qadence/ml_tools/callbacks/writer_registry.py

```diff
@@ -0,0 +1,375 @@
+from __future__ import annotations
+
+import os
+from logging import getLogger
+from types import ModuleType
+from typing import Any, Callable, Union
+from uuid import uuid4
+
+from matplotlib.figure import Figure
+from mlflow.entities import Run
+from torch import Tensor
+from torch.nn import Module
+from torch.utils.data import DataLoader
+from torch.utils.tensorboard import SummaryWriter
+
+from qadence.ml_tools.config import TrainConfig
+from qadence.ml_tools.data import DictDataLoader, OptimizeResult
+from qadence.types import ExperimentTrackingTool
+
+logger = getLogger(__name__)
+
+# Type aliases
+PlottingFunction = Callable[[Module, int], tuple[str, Figure]]
+InputData = Union[Tensor, dict[str, Tensor]]
+
+
+class BaseWriter:
+    """
+    Abstract base class for experiment tracking writers.
+
+    Methods:
+        open(config, iteration=None): Opens the writer and sets up the logging
+            environment.
+        close(): Closes the writer and finalizes any ongoing logging processes.
+        print_metrics(result): Prints metrics and loss in a formatted manner.
+        write(result): Writes the optimization results to the tracking tool.
+        log_hyperparams(hyperparams): Logs the hyperparameters to the tracking tool.
+        plot(model, iteration, plotting_functions): Logs model plots using provided
+            plotting functions.
+        log_model(model, dataloader): Logs the model and any relevant information.
+    """
+
+    run: Run  # [attr-defined]
+
+    def open(self, config: TrainConfig, iteration: int = None) -> Any:
```

Comment:
I see there are a lot of `if self.writer` checks in the other methods, which require the caller to call `open` on the writer. Is there a reason why `open` shouldn't be executed directly in the `__init__`?

From `callbackmanager.py`:
```python
writer = get_writer(tracking_tool)
writer.open(self.config, iteration=trainer.global_step)
```
from which I see that `open` is called just after initialization. I think then that we can `open` _at_ initialization.

Reply:
1) We will need to have a method to set up the writer in case the log_folder changes during runtime. For example when create_subfolders_per_run = True, a new writer will have to be set up for each run (based on a new log_dir), or we will have to modify the existing writer - both of which will need a writer opening method at the start of training. 

2) I wanted to keep the init for the writers as clean as possible - so that the get_writer function only has one argument - which is the tracking tool - and returns the object of a particular writer. After which we can open the writer using TrainConfig and iterations - the user can also open the writer with a different Config, etc. 
This might be useful if the user wants to train - without Trainer - but still want to use our writers.

```
writer = get_writer(tracking_tool)
writer.open(self.config, iteration=trainer.global_step)

writer.close()

writer.open(Differfent_config, iteration=trainer.global_step)
writer.close()
```

Reply:
Sure, sounds good.